### PR TITLE
Exercises

### PR DESCRIPTION
**Pull request by MananKumawat**

### CODE REVIEW

File: Day2/ex1.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			frequency.mutex.Unlock()
+	}
+	fmt.Println(word," is done by worker ", id)
+
+}
+
+
+func main(){
+	words := []string{"quick", "dog", "dog", "dog", "dog"}
+	numWords := len(words)
```
Comment (zahlekhan) [inline Day2/ex1.go]: move this to line 36 for better code organization

File: Day2/ex1.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+func (frequency *Frequency) FrequencyCalculator(id int, jobs <-chan string, wg *sync.WaitGroup)  { // go routine for calculating frequency of a word
+
+	defer wg.Done()
+	word := <- jobs
+
+	for _, letter := range word	{
+			frequency.mutex.Lock()
+			frequency.frequency[string(letter)] += 1
+			frequency.mutex.Unlock()
+	}
```
Comment (zahlekhan) [inline Day2/ex1.go:23]: numOfWorkers are coupled with numOfWords

if numOfWorkers < numOfWords then deadlock occurs at line 42
if numOfWorkers > numOfWords then subset of subroutines get blocked indefinitely

```suggestion
	
     for word := range jobs {
	for _, letter := range word	{
			frequency.mutex.Lock()
			frequency.frequency[string(letter)] += 1
			frequency.mutex.Unlock()
	}
}
```

File: Day2/ex2.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	totalrating.mutex.Lock()
+	totalrating.TotalRating += 0.01 * float64(rand.Intn(1000))
+	totalrating.mutex.Unlock()
+}
+
+func main() {
+	rand.Seed(time.Now().UnixNano()) // Changing seed according to time
+	const NumberOfStudents = 200
+	var totalrating = Rating{0.0, sync.Mutex{}}
+	Students := []Student{}
```
Comment (zahlekhan) [inline Day2/ex2.go]: using var is more idiomatic

```suggestion
	var Students []Student
```

File: Day2/ex2.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	var wg sync.WaitGroup
+
+	for StudentId :=0; StudentId<NumberOfStudents; StudentId++ {
+		Students = append(Students, Student{StudentId})
+	}
+	for _, student := range Students {
+		wg.Add(1)
+		go student.getrating(&wg, &totalrating)
+	}
+	wg.Wait()
```
Comment (zahlekhan) [inline Day2/ex2.go:46]: newline between 43 and 44

File: Day2/ex3.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	bankaccount.balance -= amount
+	bankaccount.mutex.Unlock()
+	fmt.Println("Current Balance:", bankaccount.balance)
+}
+
+func main() {
+	rand.Seed(time.Now().UnixNano())
+	bankaccount := BankAccount{"2030xxxx", 500, sync.Mutex{}} // Initializing a bank account with 500 Rs.
+
+	fmt.Println("Initial Banalce:", bankaccount.balance)
```
Comment (zahlekhan) [inline Day2/ex3.go]: Balance typo

File: Day2/ex3.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		amount := rand.Intn(20) * 100
+		if rand.Intn(2) == 0 {
+			wg.Add(1)
+			go bankaccount.AddMoney(amount, &wg)
+		} else {
+			wg.Add(1)
+			go bankaccount.SubtractMoney(amount, &wg)
+		}
+	}
+
```
Comment (zahlekhan) [inline Day2/ex3.go:72]: Add a method for getting balance

**Comment by MananKumawat** [comment]
Done with the mentioned changes

**Pull request merged**